File: PostProcessing/lua/Runtime/Effects/FxaaRenderer.lua
FxaaRenderer = ScriptableObject(PostProcessEffectRenderer)

-- uniform names
local u_textureSize = "u_texture_size"

-- pass ids
local FxaaPass = 0

function FxaaRenderer : ctor()
end

function FxaaRenderer : render(sys, renderContext)
    local enable = self.settings["pFxaaEnable"]
    local commands = renderContext:getCommandBuffer()
    local width = renderContext:getScreenWidth()
    local height = renderContext:getScreenHeight()

    if enable then
        local w = width
        local h = height
        local src = renderContext:getSource()
        local dst = renderContext:getDestination()
        local fxaaMat = renderContext:getResources():getShaders("Fxaa"):getMaterial()
        fxaaMat:setVec2(u_textureSize, Amaz.Vector2f(width, height))
        if self.dirty then
            if src.image == dst.image then
                local pingpong = CreateRenderTexture("", w, h)
                commands:blitWithMaterial(src, pingpong, fxaaMat, FxaaPass)
                commands:blit(pingpong, dst)
            else
                commands:blitWithMaterial(src, dst, fxaaMat, FxaaPass)
            end
            self.dirty = false
        end
    else
        if self.dirty then
            commands:clearAll()
            self.dirty = false
        end
    end

    local cam = renderContext:getCamera()
    cam.entity.scene:commitCommandBuffer(commands)
end

File: PostProcessing/js/Runtime/Effects/FxaaRenderer.js
const { Utils } = require("PostProcessing/Base/Utils");
const { PostProcessBaseRenderer } = require("PostProcessing/Runtime/Effects/PostProcessBaseRenderer");
const Amaz = effect.Amaz;

// uniform names
const u_textureSize = "u_texture_size"

// pass ids
const FxaaPass = 0

class FxaaRenderer extends PostProcessBaseRenderer{
    constructor()
    {
        super();
        this.commands = new Amaz.CommandBuffer();

        this.setRendererProperty("pFxaaFastMode", 1);
        this.setRendererProperty("pFxaaKeepAlpha", 1);
        this.setRendererProperty("pFxaaEnable", false, true);
    }

    update(properties)
    {
        this.updateProperties(properties);
    }

    render(scene, postProcessContext)
    {
        var enable = this.settings.get("pFxaaEnable");
        var width = postProcessContext.getScreenWidth();
        var height = postProcessContext.getScreenHeight();

        if ( enable )
        {
            var w = width;
            var h = height;
            var src = postProcessContext.getSource();
            var dst = postProcessContext.getDestination();
            var fxaaMat = postProcessContext.getResources().getShaders("Fxaa").getMaterial();
            
            fxaaMat.setVec2(u_textureSize, new Amaz.Vector2f(width, height));

            if ( this.dirty )
            {
                if ( src.image === dst.image )
                {
                    var pingpong = Utils.CreateRenderTexture("", w, h);
                    this.commands.blitWithMaterial(src, pingpong, fxaaMat, FxaaPass);
                    this.commands.blit(pingpong, dst);
                }else{
                    this.commands.blitWithMaterial(src, dst, fxaaMat, FxaaPass);
                }            
                this.dirty = false;
            }    
        }
        else if ( this.dirty )
        {
            this.commands.clearAll();
            this.dirty = false;
        }
        var cam = postProcessContext.getCamera();
        cam.entity.scene.commitCommandBuffer(this.commands);
    }
}

exports.FxaaRenderer = FxaaRenderer;